x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -130,6 +134,23 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeMap(metadata);
     }
 
+    public boolean sameUserAs(Authentication other) {
+        if (AuthenticationType.API_KEY == getAuthenticationType() && AuthenticationType.API_KEY == other.getAuthenticationType()) {
+            assert getUser().principal().equals(other.getUser().principal()) :
+                "The same API key ID cannot be attributed to two different usernames";
+            return getMetadata().get(API_KEY_ID_KEY).equals(other.getMetadata().get(API_KEY_ID_KEY));
+        }
+        if (false == getUser().principal().equals(other.getUser().principal())) {
+            return false;
+        }
+        final RealmRef thisRealm = getSourceRealm();
+        final RealmRef otherRealm = other.getSourceRealm();
+        if (FileRealmSettings.TYPE.equals(thisRealm.getType()) || NativeRealmSettings.TYPE.equals(thisRealm.getType())) {
+            return thisRealm.getType().equals(otherRealm.getType());
+        }
```

Comment:
True. I thought about `reserved` and like you said, it can already be handled by the existing logic. However I forgot about `kerberos`. Thanks!
I think it might be worthwhile to formalise the concept of "singleton" realm. This property is useful in serveral places, e.g. `Realms#init`. It can help avoid enumerating the singleton realms manually. Another potential benefit is to allow a custom realm to claim it should be a singleton and get the same behaviour as builtin singleton realms.

Reply:
> I think it might be worthwhile to formalise the concept of "singleton" realm. This property is useful in serveral places, e.g. Realms#init. It can help avoid enumerating the singleton realms manually. Another potential benefit is to allow a custom realm to claim it should be a singleton and get the same behaviour as builtin singleton realms. 

++

Reply:
I disagree on Kerberos. It's true there can be only 1, but it can be configured. That means that 2 realms with different names might represent different users.
Including it in the "singleton" check means that you cannot never replace your Kerberos realm if there is the chance that you would end up with 2 users with the same principal across the old/new.

Reply:
This is a subtle but important point. Thanks! Even though they are all of singleton type, file and native realms have stronger constraint on the singleton nature than kerboros. The difference mainly lies to their backends which always remain the same for file and native realms. I updated the code accordingly.
PS: We could use another word besides "singleton" to differentiate these two categories, any ideas?